engine/table/src/main/java/io/deephaven/engine/table/impl/UpdateByBucketHelper.java

```diff
@@ -0,0 +1,336 @@
+            restampRemovals = upstream.removed().union(upstream.getModifiedPreShift());
+        } else {
+            restampAdditions = upstream.added();
+            restampRemovals = upstream.removed();
+        }
+
+        // removes
+        if (restampRemovals.isNonempty()) {
+            final int size = (int) Math.min(restampRemovals.size(), 4096);
+            try (final RowSequence.Iterator it = restampRemovals.getRowSequenceIterator();
+                    final ChunkSource.GetContext context = timestampColumnSource.makeGetContext(size);
+                    final WritableLongChunk<? extends Values> ssaValues = WritableLongChunk.makeWritableChunk(size);
+                    final WritableLongChunk<OrderedRowKeys> ssaKeys = WritableLongChunk.makeWritableChunk(size)) {
+
+                MutableLong lastTimestamp = new MutableLong(NULL_LONG);
+                while (it.hasMore()) {
+                    RowSequence chunkRs = it.getNextRowSequenceWithLength(4096);
+
+                    // get the chunks for values and keys
+                    LongChunk<? extends Values> valuesChunk =
+                            timestampColumnSource.getPrevChunk(context, chunkRs).asLongChunk();
+                    LongChunk<OrderedRowKeys> keysChunk = chunkRs.asRowKeyChunk();
+
+                    // push only non-null values/keys into the Ssa
+                    fillChunkWithNonNull(keysChunk, valuesChunk, ssaKeys, ssaValues, lastTimestamp);
+                    timestampSsa.remove(ssaValues, ssaKeys);
+                }
+            }
+        }
+
+        // shifts
+        if (upstream.shifted().nonempty()) {
+            final int size = Math.max(
+                    upstream.modified().intSize() + Math.max(upstream.added().intSize(), upstream.removed().intSize()),
+                    (int) upstream.shifted().getEffectiveSize());
+            try (final RowSet fullPrevRowSet = source.getRowSet().copyPrev();
+                    final WritableRowSet previousToShift = fullPrevRowSet.minus(restampRemovals);
+                    final ColumnSource.GetContext getContext = timestampColumnSource.makeGetContext(size)) {
+
+                // no need to consider upstream removals
+                previousToShift.remove(upstream.removed());
+
+                final RowSetShiftData.Iterator sit = upstream.shifted().applyIterator();
+                while (sit.hasNext()) {
+                    sit.next();
+                    try (final RowSet subRowSet = previousToShift.subSetByKeyRange(sit.beginRange(), sit.endRange())) {
+                        if (subRowSet.isEmpty()) {
+                            continue;
+                        }
+
+                        final LongChunk<? extends Values> shiftValues =
+                                timestampColumnSource.getPrevChunk(getContext, subRowSet).asLongChunk();
+
+                        timestampSsa.applyShiftReverse(shiftValues, subRowSet.asRowKeyChunk(), sit.shiftDelta());
```

Comment:
Why do you only have to apply one direction? I'm not extremely familiar with the SSA code, but I see AJ applying in two different directions based on polarity.

Reply:
This seems really obvious and it surprises me that this isn't causing errors.  I have implemented the polarity check.